### PR TITLE
Feature repeatable text generation

### PR DESCRIPTION
**Pull request by ronanstokes-db**
## Proposed changes
Enhance text generation from Template Text Generator

Switch to vectorized implementation using Numpy / Pandas and take advantage of Numpy random number generation repeatability

This PR implements repeatable vectorized implementation of TemplateTextGenerator

It also expands coverage of text generators substantially

## Types of changes

What types of changes does your code introduce to dbldatagen?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Change to tutorials, tests or examples
- [ ] Non code change (readme, images or other non-code assets)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. 
If you're unsure about any of them, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Submission does not reduce code coverage numbers
- [x] Submission does not increase alerts or messages from LGTM

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you 
did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by howardwu-db** [comment]
sample comment

File: dbldatagen/text_generators.py [hunk 1]
```diff
@@ -177,24 +177,24 @@ class TemplateGenerator(TextGenerator):  # lgtm [py/missing-equals]
...
-    A         Insert a random uppercase alphabetical character
-    k         Insert a random lowercase alphanumeric character
-    K         Insert a random uppercase alphanumeric character
-    n         Insert a random number between 0 .. 255 inclusive. This option must always be escaped
-    N         Insert a random number between 0 .. 65535 inclusive. This option must always be escaped
-    w         Insert a random lowercase word from the ipsum lorem word set. Always escaped
-    W         Insert a random uppercase word from the ipsum lorem word set. Always escaped
-    ========  ======================================
+    v0,v1,..v9 Use base value as an array of values and substitute the `nth` element ( 0 .. 9). Always escaped.
+    x          Insert a random lowercase hex digit
```
Comment (howardwu-db) [inline dbldatagen/text_generators.py:185]: is there a reason why we use x for hex? I feel like it's more universal to represent with 'h' but I do know that "0x" is typical for hex

Comment (howardwu-db) [reply dbldatagen/text_generators.py:185]: ah nevermind I see that python itself uses x for hex https://docs.python.org/3/library/string.html#format-specification-mini-language

**Review by howardwu-db — CHANGES_REQUESTED** [review]
Have 3 questions

File: dbldatagen/text_generators.py [hunk 2]
```diff
@@ -259,15 +279,20 @@ def _getRandomInt(self, low, high=-1, rng=None):
         # use standard random for now as it performs better when generating values one at a time
         return random.randint(low, high)
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/text_generators.py:280]: performs better

File: dbldatagen/text_generators.py [hunk 3]
```diff
@@ -289,71 +315,241 @@ def stringsFromSingleTemplate(self, baseValue, genTemplate, escapeSpecialMeaning
...
+        # in the following code, the construct `(not escape) ^ self._escapeSpecialMeaning` means apply
+        # special meaning if either escape is not true or the option `self._escapeSpecialMeaning` is true.
+        # This corresponds to the logical xor operation
+        for i in range(0, template_len):
+            char = genTemplate[i]
+            following_char = genTemplate[i + 1] if i + 1 < template_len else None
+
+            if char == '\\':
+                escape = True
+            elif use_value and ('0' <= char <= '9'):
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/text_generators.py:456]: there is a different use in the if case so not every rule is the same

File: dbldatagen/text_generators.py [hunk 4]
```diff
@@ -362,23 +558,104 @@ def classicGenerateText(self, v):
...
+        placeholders = np.full((v.shape[0], self._max_placeholders), '', dtype=np.object_)
+
+        # prepare template selections, bounds, rnd values to drive application of algorithm
+        template_choices, template_rnd_bounds, template_rnds = self._prepare_random_bounds(v)
+        template_choices_t = template_choices.T
+
+        # create masked arrays, with all elements initially masked
+        # as we substitute template expansion, we'll mask and unmask rows corresponding to each template
+        # calling the method to substitute the values on the masked placeholders
+        masked_placeholders = np.ma.MaskedArray(placeholders, mask=False)
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/text_generators.py:625]: yes - intent is to start unmasked and mask rows as needed

**Review by howardwu-db — APPROVED** [review]

**Pull request merged**